libs/vectorise/include/vectorise/uint/int.hpp

```diff
@@ -0,0 +1,1298 @@
+  for (std::size_t i = 0; i < WIDE_ELEMENTS; ++i)
+  {
+    // if sum of elements is smaller than the element itself, then we have overflow and carry
+    new_carry = (wide_[i] + n.ElementAt(i) + carry < wide_[i]) ? 1 : 0;
+    wide_[i] += n.ElementAt(i) + carry;
+    carry = new_carry;
+  }
+
+  return *this;
+}
+
+template <uint16_t S>
+constexpr Int<S> &Int<S>::operator-=(Int<S> const &n)
+{
+  if (n < _0)
+  {
+    *this += -n;
+    return *this;
+  }
+  Int<S>::WideType carry = 0, new_carry = 0;
+  for (std::size_t i = 0; i < WIDE_ELEMENTS; ++i)
+  {
+    // if diff of the elements is larger than the element itself, then we have underflow and carry
+    new_carry = (wide_[i] < n.ElementAt(i) + new_carry) ? 1 : 0;
+    wide_[i] -= n.ElementAt(i) + carry;
+    carry = new_carry;
+  }
+
+  return *this;
+}
+
+// Implementation for 256-bits only
+template <>
+constexpr Int<256> &Int<256>::operator*=(Int<256> const &n)
+{
+  bool sign = true;
+  if (*this < _0)
+  {
+    *this = -*this;
+    sign  = !sign;
+  }
+  Int<256> o{n};
+  if (o < _0)
+  {
+    o    = -o;
+    sign = !sign;
+  }
+
+  // Calculate all products between each uint64_t element in the Ints
+  // Use int128_t type to hold the actual product.
+  __uint128_t products[WIDE_ELEMENTS][WIDE_ELEMENTS] = {};
+
+  for (std::size_t i = 0; i < WIDE_ELEMENTS; ++i)
+  {
+    for (std::size_t j = 0; j < WIDE_ELEMENTS; ++j)
+    {
+      // Note: C++14 does not have constexpr std::array, we need to cast the array
+      products[i][j] =
+          static_cast<__uint128_t>(wide_[i]) * static_cast<__uint128_t>(o.ElementAt(j));
+    }
+  }
+
+  /* If a, b, two Int<256> numbers, with the following elements (little endian order):
+   * |  a[3]  |  a[2]  |  a[1]  |  a[0]  |
+   * |  b[3]  |  b[2]  |  b[1]  |  b[0]  |
+   * Then both can be written in the following form:
+   * a = a[0] + (a[1] << 64) + (a[2] << 128) + (a[3] << 192)
+   * b = b[0] + (b[1] << 64) + (b[2] << 128) + (b[3] << 192)
+   *
+   * Then a * b is the following, truncating terms left shifted over 192 bits:
+   * a * b =  a[0] * b[0]
+   *       + (a[0] * b[1] + a[1] * b[0]) << 64
+   *       + (a[0] * b[2] + a[1] * b[1] + a[2] * b[0]) << 128
+   *       + (a[0] * b[3] + a[1] * b[2] + a[2] * b[1] + a[3] * b[0]) << 192
+   * In order to be accurate, we calculate these terms as 128-bit quantities and we add the
+   * high-64-bits as carry to the next term
+   */
+
+  __uint128_t carry = 0, terms[WIDE_ELEMENTS] = {};
+  terms[0] = products[0][0];
+  carry    = static_cast<WideType>(terms[0] >> WIDE_ELEMENT_SIZE);
+  terms[1] = products[0][1] + products[1][0] + carry;
+  carry    = static_cast<WideType>(terms[1] >> WIDE_ELEMENT_SIZE);
+  terms[2] = products[0][2] + products[1][1] + products[2][0] + carry;
+  carry    = static_cast<WideType>(terms[2] >> WIDE_ELEMENT_SIZE);
+  terms[3] = products[0][3] + products[1][2] + products[2][1] + products[3][0] + carry;
+  // TODO(?): decide what to do with overflow if carry > 0
```

Comment:
leftover todo
please either raise an issue or remove if redundant

Reply:
removed, overflow is taken care of in the FixedPoint class that uses Int.